pandora-client-web/src/components/header/leaveButton.tsx

```diff
@@ -0,0 +1,139 @@
+import { HeaderButton } from './HeaderButton';
+import { useShardConnectionInfo } from '../gameContext/shardConnectorContextProvider';
+import { usePlayer, usePlayerData, usePlayerState } from '../gameContext/playerContextProvider';
+import { useCurrentAccount, useDirectoryConnector } from '../gameContext/directoryConnectorContextProvider';
+import { EMPTY, GetLogger, IChatRoomFullInfo } from 'pandora-common';
+import { Button } from '../common/button/button';
+import { useLogout } from '../../networking/account_manager';
+import { useCharacterRestrictionsManager, useChatRoomInfo } from '../gameContext/chatRoomContextProvider';
+import { PlayerCharacter } from '../../character/player';
+import { toast } from 'react-toastify';
+import { TOAST_OPTIONS_ERROR } from '../../persistentToast';
+import { ModalDialog } from '../dialog/dialog';
+import { Column } from '../common/container/container';
+import './leaveButton.scss';
+import { HoverElement } from '../hoverElement/hoverElement';
+
+const leaveButtonContext = createContext(() => { /** noop */ });
+
+export function LeaveButton() {
+	const [show, setShow] = useState(false);
+	const closeDialog = useCallback(() => setShow(false), []);
+
+	return (
+		<leaveButtonContext.Provider value={ closeDialog }>
+			<HeaderButton icon={ logoutIcon } iconAlt='Leave' onClick={ () => setShow(true) } title='Leave' />
+			{
+				show ? <DialogLeave /> : null
+			}
+		</leaveButtonContext.Provider>
+	);
+}
+
+function DialogLeave(): ReactElement {
+	const closeDialog = useContext(leaveButtonContext);
+
+	return (
+		<ModalDialog>
+			<Column className='LeaveDialog'>
+				<ChatRoomLeave />
+				<CharacterLeave />
+				<AccountLeave />
+				<Button onClick={ closeDialog }>Cancel</Button>
+			</Column>
+		</ModalDialog>
+	);
+}
+
+function ChatRoomLeave(): ReactElement | null {
+	const player = usePlayer();
+	const room = useChatRoomInfo();
+
+	if (!player || !room) {
+		return null;
+	}
+
+	return (
+		<CharRoomLeaveInner player={ player } room={ room } />
+	);
+}
+
+function CharRoomLeaveInner({ player, room }: { player: PlayerCharacter; room: IChatRoomFullInfo; }): ReactElement | null {
+	const directoryConnector = useDirectoryConnector();
+	const playerState = usePlayerState();
+	const canLeave = useCharacterRestrictionsManager(playerState, player, (manager) => (manager.isInSafemode() || !manager.getEffects().blockRoomLeave));
+	const [leaveButtonRef, setLeaveButtonRef] = useState<HTMLButtonElement | null>(null);
+
+	const onRoomLeave = useCallback(() => {
+		directoryConnector.awaitResponse('chatRoomLeave', EMPTY)
+			.then((result) => {
+				if (result.result !== 'ok') {
+					toast(`Failed to leave room:\n${result.result}`, TOAST_OPTIONS_ERROR);
+				}
+			})
+			.catch((err) => {
+				GetLogger('LeaveRoom').warning('Error during room leave', err);
+				toast(`Error during room creation:\n${err instanceof Error ? err.message : String(err)}`, TOAST_OPTIONS_ERROR);
```

Comment:
Likely already was that way before, but should be fixed anyway:
```suggestion
				toast(`Error while leaving room:\n${err instanceof Error ? err.message : String(err)}`, TOAST_OPTIONS_ERROR);
```